core/src/test/java/org/elasticsearch/index/seqno/CheckpointsIT.java

```diff
@@ -60,17 +60,17 @@ public void testCheckpointsAdvance() throws Exception {
                 logger.debug("seq_no stats for {}: {}", shardStats.getShardRouting(),
                     XContentHelper.toString(shardStats.getSeqNoStats(),
                         new ToXContent.MapParams(Collections.singletonMap("pretty", "false"))));
-                final Matcher<Long> localCheckpointRule;
+                final Matcher<Long> globalCheckpointRule;
                 if (shardStats.getShardRouting().primary()) {
-                    localCheckpointRule = equalTo(numDocs - 1L);
+                    globalCheckpointRule = equalTo(numDocs - 1L);
                 } else {
-                    // nocommit:  recovery doesn't transfer local checkpoints yet (we don't persist them in lucene).
-                    localCheckpointRule = anyOf(equalTo(numDocs - 1L), equalTo(SequenceNumbersService.NO_OPS_PERFORMED));
+                    // nocommit: recovery doesn't transfer global checkpoints yet (we don't persist them in lucene).
+                    globalCheckpointRule = anyOf(equalTo(numDocs - 1L), equalTo(SequenceNumbersService.UNASSIGNED_SEQ_NO));
                 }
                 assertThat(shardStats.getShardRouting() + " local checkpoint mismatch",
-                    shardStats.getSeqNoStats().getLocalCheckpoint(), localCheckpointRule);
+                    shardStats.getSeqNoStats().getLocalCheckpoint(), equalTo(numDocs - 1L));
```

Comment:
> if recovery completes after all docs were indexed, we will not have a local checkpoint on the replica?

Correct, if there was no translog replay component. In this case, no indexing operations will have been performed on the recovery target and the local checkpoint there will not have advanced.

Note that the logs [here](https://gist.github.com/jasontedor/0e25d9dbcb49f706e98c4cd83b210551) are for a case where the recovery completes after the docs were indexed, but there *was* a translog replay.

I pushed 2545a351326740307ce538290a43b59cad825b73.